core/emulate.c

```diff
@@ -920,17 +987,63 @@ static void em_movsx(struct em_context_t *ctxt)
     ctxt->dst.value = 0;
     memcpy(&ctxt->dst.value, &extension, ctxt->dst.size);
     memcpy(&ctxt->dst.value, &ctxt->src1.value, ctxt->src1.size);
+    return EM_CONTINUE;
+}
+
+static em_status_t em_push(struct em_context_t *ctxt)
+{
+    em_status_t rc;
+    em_operand_t sp;
+
+    // Pre-decrement stack pointer
+    register_add(ctxt, REG_RSP, -(int64_t)ctxt->operand_size);
+
+    memset(&sp, 0, sizeof(sp));
+    sp.type = OP_MEM;
+    sp.size = ctxt->operand_size;
+    sp.mem.ea = gpr_read(ctxt, REG_RSP, ctxt->address_size);
+    sp.mem.seg = SEG_SS;
+    sp.value = ctxt->src1.value;
+
+    rc = operand_write(ctxt, &sp);
+    return rc;
+}
+
+static em_status_t em_pop(struct em_context_t *ctxt)
+{
+    em_status_t rc;
+    em_operand_t sp;
+
+    memset(&sp, 0, sizeof(sp));
+    sp.type = OP_MEM;
+    sp.size = ctxt->operand_size;
+    sp.mem.ea = gpr_read(ctxt, REG_RSP, ctxt->address_size);
```

Comment:
There's a detail in the PUSH/POP spec about which \*SP register (SP, ESP or RSP) should be used, i.e. the stack-address size, which can be different from `ctxt->address_size`:

> Stack-address size. Outside of 64-bit mode, the B flag in the current stack-segment descriptor determines the
size of the stack pointer (16 or 32 bits); in 64-bit mode, the size of the stack pointer is always 64 bits.
> The stack-address size determines the width of the stack pointer when reading from the stack in memory and
when incrementing the stack pointer. (As stated above, the amount by which the stack pointer is incremented
is determined by the operand size.)

To be fully compliant, I believe we need to check the B (Big) bit of the VMCS `GUEST_SS_AR` field (cf. Intel SDM Vol. 3A, Section 3.4.5: _Segment Descriptors_). If that requires too much additional effort, please leave a TODO comment instead.

Reply:
Nice catch, thanks a lot for your feedback! I wasn't aware of this. I'll place a *TODO* for now, since it requires rather big changes to the vcpu-emulator interface.